src/Architecture.cpp

```diff
@@ -687,3 +687,52 @@ void Architecture::printCouplingMap(const CouplingMap& cm, std::ostream& os) {
   }
   os << "}" << std::endl;
 }
+
+GateFlipStrategy Architecture::getGateFlipStrategy(qc::OpType opType) {
+  switch (opType) {
+  case qc::None:
+    return GateFlipStrategy::Unknown;
+  case qc::X:
+  case qc::SX:
+  case qc::SXdag:
+    return GateFlipStrategy::Hadamard;
+  case qc::Z:
+  case qc::S:
+  case qc::Sdag:
+  case qc::T:
+  case qc::Tdag:
+  case qc::Phase:
+  case qc::RZ:
+  case qc::SWAP:
+    return GateFlipStrategy::Identity;
+  default:
+    return GateFlipStrategy::Swap;
+  }
+}
+
+std::uint32_t Architecture::computeCostDirectionReverse(qc::OpType opType) {
+  switch (Architecture::getGateFlipStrategy(opType)) {
+  case GateFlipStrategy::Identity:
+    return 0;
+  case GateFlipStrategy::Hadamard:
+    return 4 * COST_SINGLE_QUBIT_GATE;
+  case GateFlipStrategy::Swap:
+  case GateFlipStrategy::Unknown:
+    return 2 * COST_BIDIRECTIONAL_SWAP;
+  default:
+    throw QMAPException("Unexpected GateFlipStrategy!");
+  }
```

Comment:
The reason for Unknown was mainly with the distanceTable in mind, because it doesn't have gate context. In that case we still need some cost though.

Reply:
Just a guess, but I could imagine that it could be worth it to add two more distance tables for an architecture that is uni-directional. Then one would have all the cases covered.
I think it could be important to only add this overhead on directed architectures.